### ISSUE TITLE
no grad for  TexturesVertex   

### ISSUE THREAD

**Issue by Jiakui** [facebookresearch/pytorch3d#763]
==========================================

# Load the obj and ignore the textures and materials.
verts, faces_idx, _ = load_obj("./data/lg_m.obj")
faces = faces_idx.verts_idx

verts_shape = verts.shape 
deform_verts = torch.full(verts_shape, 0.0, device=device, requires_grad=True)  
sphere_verts_rgb = torch.full([1, verts_shape[0], 3], 0.8, device=device ,  requires_grad=True) 

# Create a Meshes object for the teapot. Here we have only one mesh in the batch.
leg_mesh = Meshes(
    verts=[verts.to(device)],   
    faces=[faces.to(device)], 
    textures=TexturesVertex(verts_features=sphere_verts_rgb) 
)

==============================================

ABOVE is the code   

during optimization, I print  the grad by  **print(sphere_verts_rgb.grad_fn)** , and get **None**.  

The pytorch3d I use is  0.4.0, with pytorch 1.7.1 .

Thanks!

**Comment by Jiakui** [comment]
no grad for   TexturesVertex,   and then I cannot  optimize for the texture of  mesh.   Please help help help !!!!!!!!!!!!!!!!!!!!!!

**Comment by Jiakui** [comment]
@nikhilaravi

**Comment by Jiakui** [comment]
@gkioxari

**Comment by bottler** [comment]
Can you paste or summarize the rest of your code, especially the model? If we know what you are optimizing, we might be able to help.

**Comment by Jiakui** [comment]
num_views = 4 
K = torch.load("./data/camera_params/K.pth")
R = torch.load("./data/camera_params/R.pth")
T = torch.load("./data/camera_params/T.pth")
image_size = torch.load("./data/camera_params/image_size.pth")
target_cameras = PerspectiveCameras(device=device, K=K, R=R, T=T, image_size=image_size)
#pdb.set_trace() 

# Rasterization settings for differentiable rendering, where the blur_radius
# initialization is based on Liu et al, 'Soft Rasterizer: A Differentiable 
# Renderer for Image-based 3D Reasoning', ICCV 2019
sigma = 1e-4

raster_settings_soft = RasterizationSettings(
    image_size=(180, 135), 
    blur_radius=np.log(1. / 1e-4 - 1.)*sigma, 
    faces_per_pixel=50, 
    perspective_correct=False
)

lights = PointLights(device=device, location=[[0.0, 0.0, -3.0]])
#lights = AmbientLights(device=device,  )

blend_params = BlendParams(background_color=(0., 0., 0. ) ) 

# Differentiable soft renderer using per vertex RGB colors for texture
renderer_textured = MeshRenderer(
    rasterizer=MeshRasterizer(
        cameras=target_cameras, 
        raster_settings=raster_settings_soft
    ),
    shader=SoftPhongShader(device=device, 
        cameras=target_cameras,
        lights=lights,
        blend_params=blend_params 
        )
)


image_ref = []


#pdb.set_trace() 
target_silhouette = torch.from_numpy(np.array(image_ref )).to(device) 


image_ref = []


#pdb.set_trace() 
target_rgb = torch.from_numpy(np.array(image_ref )).to(device) 



# Number of views to optimize over in each SGD iteration
num_views_per_iteration = 2
# Number of optimization steps
Niter = 2000
# Plot period for the losses
plot_period = 250


# Optimize using rendered silhouette image loss, mesh edge loss, mesh normal 
# consistency, and mesh laplacian smoothing
losses = {  "rgb": {"weight": 100.0, "values": []},  
            "silhouette": {"weight": 100.0, "values": []},
          "edge": {"weight": 1.0, "values": []},
          "normal": {"weight": 1.0, "values": []},
          "laplacian": {"weight": 1.0, "values": []},
          "l2": {"weight": 1.0, "values": []},
          #"rgb_l2": {"weight": 1.0, "values": []},
          "norm_loss": {"weight": 10.0, "values": []},
         }

# Losses to smooth / regularize the mesh shape
def update_mesh_shape_prior_losses(mesh, loss):
    # and (b) the edge length of the predicted mesh
    loss["edge"] = mesh_edge_loss(mesh)
    
    # mesh normal consistency
    loss["normal"] = mesh_normal_consistency(mesh)[0]
    
    # mesh laplacian smoothing
    loss["laplacian"] = mesh_laplacian_smoothing(mesh, method="uniform")


# Load the obj and ignore the textures and materials.
verts, faces_idx, _ = load_obj("./data/lg_m.obj")
faces = faces_idx.verts_idx

# Initialize each vertex to be white in color.
#verts_rgb = torch.ones_like(verts)[None]  # (1, V, 3)
#textures = TexturesVertex(verts_features=verts_rgb.to(device))

#verts_shape = leg_mesh.verts_packed().shape
verts_shape = verts.shape 
deform_verts = torch.full(verts_shape, 0.0, device=device, requires_grad=True)  
sphere_verts_rgb = torch.full([1, verts_shape[0], 3], 0.8, device=device ,  requires_grad=True) 
#sphere_verts_rgb  -=  torch.tensor([0.8, 0.2, 0.0], device=device ) 


# Create a Meshes object for the teapot. Here we have only one mesh in the batch.
leg_mesh = Meshes(
    verts=[verts.to(device)],   
    faces=[faces.to(device)], 
    textures=TexturesVertex(verts_features=sphere_verts_rgb) 
)

# We will learn to deform the source mesh by offsetting its vertices
# The shape of the deform parameters is equal to the total number of vertices in
# src_mesh

rotation =  torch.tensor([-0.0251,  0.0039,  0.0444], dtype=torch.float32, device=device )  
translation =  torch.tensor([[0.0686, 0.1402, 0.0308]], dtype=torch.float32, device=device )    
scale =  torch.tensor([1.0446], dtype=torch.float32, device=device )   


# The optimizer
optimizer = torch.optim.SGD([ deform_verts, sphere_verts_rgb], lr=0.0001, momentum=0.9)
scheduler = StepLR(optimizer, step_size=100, gamma=0.1)


#loop = tqdm(range(Niter))
n1 = leg_mesh.verts_normals_padded()[:, leg_mesh.faces_padded()[0], :].clone()
n1 = torch.mean(n1, dim=2)


def huber(x, y, scaling=0.1):
    diff_sq = (x - y) ** 2
    loss = ((1 + diff_sq / (scaling**2)).clamp(1e-4).sqrt() - 1) * float(scaling)
    return loss


for i in range(60):
    # Initialize optimizer
    optimizer.zero_grad()
    
    #leg_mesh.textures = TexturesVertex(verts_features=sphere_verts_rgb) 
    # Deform the mesh
    new_src_mesh = leg_mesh.offset_verts(deform_verts)
    

    """Veily add Norm_loss"""
    n2 = new_src_mesh.verts_normals_padded()[:, new_src_mesh.faces_padded()[0], :]
    #pdb.set_trace()
    n2 = torch.mean(n2, dim=2)
    #print(n1.size(), n2.size() )
    norm_loss = 1 - torch.cosine_similarity(n1, n2, dim=2)
    norm_loss = torch.mean(norm_loss)

    # Losses to smooth /regularize the mesh shape
    loss = {k: torch.tensor(0.0, device=device) for k in losses}
    update_mesh_shape_prior_losses(new_src_mesh, loss)
    loss["l2"] = torch.mean(deform_verts **2) 
    #loss["rgb_l2"] = torch.mean( (sphere_verts_rgb - 1.0 ) **2) 
    loss["norm_loss"] = norm_loss 

    temp_meshes = new_src_mesh.clone() 
    verts_src = temp_meshes.verts_padded()

    """SCALE"""
    verts_src = torch.abs( scale )  * verts_src
    temp_meshes = temp_meshes.update_padded(verts_src)

    """ROT"""
    ROT = pytorch3d.transforms.Rotate(pytorch3d.transforms.euler_angles_to_matrix(rotation, "ZYX"), device=device)
    verts_src = ROT.transform_points(verts_src)
    temp_meshes = temp_meshes.update_padded(verts_src)
    
    """TRANS"""
    TRANS = pytorch3d.transforms.Translate(translation, device=device)
    verts_src = TRANS.transform_points(verts_src)
    new_src_mesh = temp_meshes.update_padded(verts_src)
        
    #new_src_mesh.textures = TexturesVertex(verts_features=sphere_verts_rgb) 
    
    # Compute the average silhouette loss over two random views, as the average 
    # squared L2 distance between the predicted silhouette and the target 
    # silhouette from our dataset
    images_predicted = renderer_textured(new_src_mesh.extend(num_views), cameras=target_cameras, lights=lights)

    for j in np.random.permutation(num_views).tolist()[:num_views_per_iteration]:    
        predicted_silhouette = images_predicted[j, ..., 3]
        #loss_silhouette = ((predicted_silhouette - target_silhouette[j]) ** 2).mean()
        loss_silhouette = huber(predicted_silhouette ,  target_silhouette[j]).abs().mean()
        #loss_silhouette = F.binary_cross_entropy(predicted_silhouette[0], target_silhouette[j]) 
        loss["silhouette"] += loss_silhouette / num_views_per_iteration
        # Squared L2 distance between the predicted RGB image and the target 
        # image from our dataset
        predicted_rgb = images_predicted[j, ..., :3]

        #loss_rgb = ((predicted_rgb - target_rgb[j]) ** 2).mean()
        loss_rgb = huber(predicted_rgb , target_rgb[j]).abs().mean()  
        loss["rgb"] += loss_rgb / num_views_per_iteration
            
    if i%10 == 0:
        cv2.imwrite("debug_rgb/" + str(i) + ".png", images_predicted[0, ..., :3].detach().cpu().numpy()*255)  
        #cv2.imwrite("debug_rgb/" + "real" + str(i) + ".png", target_rgb[0].cpu().numpy()*255 )  

    # Weighted sum of the losses
    sum_loss = torch.tensor(0.0, device=device)
    for k, l in loss.items():
        #print(k, l) 
        sum_loss += l * losses[k]["weight"]
        losses[k]["values"].append(float(l.detach().cpu()))

    print(i, sum_loss.detach().cpu().numpy(), loss["rgb"].detach().cpu().numpy(), loss["silhouette"].detach().cpu().numpy(), loss["edge"].detach().cpu().numpy() , 
        loss["normal"].detach().cpu().numpy(), loss["laplacian"].detach().cpu().numpy() , loss["l2"].detach().cpu().numpy(), loss["norm_loss"].detach().cpu().numpy() )   
    # Print the losses
    #loop.set_description("total_loss = %.6f" % sum_loss)  loss["rgb_l2"].detach().cpu().numpy(), 
    
    print(sphere_verts_rgb.grad_fn) 
        
    # Optimization step
    sum_loss.backward()
    #nn.utils.clip_grad_norm_(deform_verts, max_norm=0.2, norm_type=2)
    optimizer.step()


    scheduler.step()
    print(scheduler.get_last_lr() )

**Comment by Jiakui** [comment]
all the codes are above, except some io operation which is not shown for simplicity

**Comment by bottler** [comment]
I edited your post just to put it all in a code block, which makes it easier to read.

**Comment by bottler** [comment]
`sphere_verts_rgb` is a leaf tensor. It is a parameter of the optimization. It is created with requires_grad. That is fine. But you wouldn't expect such a tensor to have a non-None `grad_fn`. `grad_fn` is used for non-leaf tensors, which have a backward graph.

e.g.
```
a=torch.arange(5, requires_grad=True, dtype=torch.float32)
b=2*a
assert a.grad_fn is None
assert b.grad_fn is not None
```
If you want the actual gradient, use `grad` not `grad_fn`. For example:
```
b.sum().backward()
print(a.grad)
```

**Comment by Jiakui** [comment]
Thanks for your reply.  Indeed sphere_verts_rgb.grad is not None, but it is always zero. Still I  cannot optimize the value.



(Pdb) sphere_verts_rgb.grad.mean()
tensor(0., device='cuda:0')
(Pdb) sphere_verts_rgb.grad.max()
tensor(0., device='cuda:0')
(Pdb) sphere_verts_rgb.grad.min()
tensor(0., device='cuda:0')
(Pdb) sphere_verts_rgb.grad
tensor([[[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.],
         ...,
         [0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]]], device='cuda:0')

**Comment by bottler** [comment]
Can you create the texture and the mesh inside the optimization loop? Modifying the texture's data in place without telling it isn't right. deform_verts is a special function which does allow modifying a PyTorch3D object, and is an optimization. 

Try this. Remove the lines
```
leg_mesh = Meshes(
    verts=[verts.to(device)],   
    faces=[faces.to(device)], 
    textures=TexturesVertex(verts_features=sphere_verts_rgb) 
)
```
and
```
    new_src_mesh = leg_mesh.offset_verts(deform_verts)
```
and replace the latter with
```
leg_mesh = Meshes(
    verts=[verts.to(device)+deform_verts],   
    faces=[faces.to(device)], 
    textures=TexturesVertex(verts_features=sphere_verts_rgb) 
)
```

**Comment by Jiakui** [comment]
not working,  I still have :

(Pdb) sphere_verts_rgb.grad.max()
tensor(0., device='cuda:0')
(Pdb) sphere_verts_rgb.grad.min()
tensor(0., device='cuda:0')

**Comment by Jiakui** [comment]
@bottler   actually, can you just try the fit_textured_mesh tutorial?  I think if  the  fit_textured_mesh work on your computer, it might be that I am not using the correct version of pytorch3d. Otherwise, there should be a real bug about  **TexturesVertex**.

**Comment by Jiakui** [comment]
@bottler  can you try the   fit_textured_mesh   tutorial on your computer  and   print    sphere_verts_rgb.grad    ???   Thanks so much !

**Comment by bottler** [comment]
I just ran the fit_textured_mesh tutorial (from the stable branch) on colab (which uses the stable install of PyTorch3D, i.e. version 0.4.0). I had to make some modifications to it to make it work (known from other issues), namely setting `perspective_correct=False` in every RasterizationSettings, changing the two instances of `losses[k]["values"].append(l)` to `losses[k]["values"].append(float(l.detach().cpu()))`, and wrapping `predicted_images = renderer(predicted_mesh)\n"` in `with torch.no_grad():`.

Having made these changes, the tutorial ran fine. The texture is learned. I added the print at the end and it has values in.

![image](https://user-images.githubusercontent.com/669761/126319201-8f5b8b40-3807-4b61-99ba-f6b9daa3a65f.png)

**Comment by Jiakui** [comment]
no ,  it  dose not work on my computer.  I  installed pytorched by  "conda install pytorch3d -c pytorch3d", and then run the   fit_textured_mesh tutorial,   but the  sphere_verts_rgb.grad still gives me  all zero value.  maybe the conda channel  pytorch3d is problematic.  Can I download the whl file from anywhere else?

Thanks!

@bottler

**Comment by bottler** [comment]
Can you print your `conda info` and `conda list` here? The wheel and conda packages for the same version will work the same. Can you run the tutorial on both colab and your computer and compare the results?

**Comment by zhangy76** [comment]
I encounter a similar problem. I wrote a simple program for fitting 3D mesh to image by minimizing the differences between texture rendering and image. The lighting parameters have gradients. While the texture parameters, 3D mesh model parameters and camera parameters do not have gradients.

**Comment by bottler** [comment]
@zhangy76 Your experience sounds very different to this one. Please open a new issue and give full details if you haven't solved the problem.

**Comment by zhangy76** [comment]
I will open a new issue then. Thanks.

**Comment by nikhilaravi** [comment]
@Jiakui were you able to resolve this issue using @bottler's suggestions?
